mixer_prog/mixer_prog.8
.\"-
.\" Copyright (c) 2021 Christos Margiolis <christos@FreeBSD.org>
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a copy
.\" of this software and associated documentation files (the "Software"), to deal
.\" in the Software without restriction, including without limitation the rights
.\" to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
.\" copies of the Software, and to permit persons to whom the Software is
.\" furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in
.\" all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
.\" AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
.\" OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
.\" THE SOFTWARE.
.\"
.\" $FreeBSD$
.\"

.Dd June 30, 2021
.Dt mixer 8
.Os
.Sh NAME
.Nm mixer
.Nd manipulate soundcard mixer controls
.Sh SYNOPSIS
.Nm
.Op Fl f Ar device
.Op Fl d Ar unit
.Op Fl os
.Op Ar dev Ns Op . Ns Ar control Ns Op = Ns Ar value
.Ar ...
.Nm
.Op Fl d Ar unit
.Op Fl os
.Fl a
.Sh DESCRIPTION
The
.Nm
utility is used to set and display soundcard mixer device controls.
.Pp
The options are as follows:
.Bl -tag -width "-f device"
.It Fl a
Print the values for all mixer devices available in the system (see FILES).
.It Fl d Ar unit
Change the default audio card to
.Ar unit .
The unit has to be an integer value. To see what unit values are available, look
at the number each mixer device has by running
.Nm .
.It Fl f Ar device 
Open
.Ar device
as the mixer device (see FILES).
.It Fl o
Print mixer values in a format suitable for use inside scripts. The
mixer's header (name, audio card name, ...) will not be printed.
.It Fl s
Print only the recording source(s) of the mixer device.
.El
.Pp
The list of mixer devices that may be modified are:
.Bd -ragged -offset indent
vol, bass, treble, synth, pcm, speaker, line, mic, cd, mix,
pcm2, rec, igain, ogain, line1, line2, line3, dig1, dig2, dig3,
phin, phout, video, radio, and monitor.
.Ed
.Pp
Not all mixer devices are available.
.Pp
Without any arguments,
.Nm
displays all information for each one of the mixer's supported devices to
.Ar stdout .
If the
.Ar dev
argument is specified,
.Nm
displays only the values for
.Ar dev .
More than one device may be specified.
.Pp
Commands use the following format:
.Pp
.Bl -column xxxxxxxxxxxxxxxxxxxxxxxx -offset indent
.It Sy "Name	Action"
.It "dev	Display all controls"
.It "dev.control	Display only the specified control"
.It "dev.control=value	Set control value"
.El
.Pp
The available controls are as follows (replace
.Ar dev
with one of the available devices):
.Bl -column xxxxxxxxxxxxxxxxxxxxxxxx -offset indent
.It Sy "Name	Value"
.It "dev.volume	[[+|-]lvol[:[+|-]rvol]]"
.It "dev.mute	{0|1|^}"
.It "dev.recsrc	{+|-|^|=}"
.El
.Pp
The
.Ar dev.volume
control modifies a device's volume. The optional
.Ar lvol 
and/or
.Ar rvol
values have to be specified. The values have to be normalized 32-bit floats
(0.0 to 1.0). If the the left or right volume values are prefixed with
.Cm +
or
.Cm - ,
the value following will be used as a relative adjustment, modifying the
current settings by the amount specified.
.Pp
The
.Ar dev.mute
control (un)mutes a device. The following values are available:
.Bl -tag -width = -offset indent
.It Cm 0
unmutes
.Ar dev .
.It Cm 1
mutes
.Ar dev .
.It Cm ^
toggles the mute of
.Ar dev .
.El
.Pp
The
.Ar dev.recsrc
control modifies the recording sources of a mixer.
.Nm
marks devices which can be used as a recording source with
.Ar rec .
Recording sources are marked with
.Ar src .
To modify the recording source you can use one of the following modifiers
on a
.Ar rec
device:
.Bl -tag -width = -offset indent
.It Cm ^
toggles
.Ar dev
of possible recording devices
.It Cm +
adds
.Ar dev
to possible recording devices
.It Cm -
removes
.Ar dev
from possible recording devices
.It Cm =
sets the recording device to
.Ar dev
.El
.Sh FILES
.Bl -tag -width /dev/mixerN -compact
.It Pa /dev/mixerN
The mixer device, where
.Ar N
is the number of that device, for example
.Ar /dev/mixer0 .
PCM cards and mixers have a 1:1 relationship, which means that 
.Ar mixer0
is the mixer for
.Ar pcm0 
and so on. By default,
.Nm
prints both the audio card's number and the mixer associated with it
in the form of
.Ar pcmN:mixer .
The
.Ar /dev/mixer
file, although it doesn't exist in the filesystem, points to the default
mixer device and is the file
.Nm
opens when the
.Fl f Ar device
option has not been specified.
.El
.Sh EXAMPLES
.Pp
Change the volume for the
.Ar vol
device of the
.Ar /dev/mixer0
mixer device to 0.65:
.Bl -tag -width Ds -offset indent
.It $ mixer -f /dev/mixer0 vol.volume=0.65
.El
.Pp
Increase the 
.Ar mic
device's left volume by 0.10 and decrease the right
volume by 0.05:
.Bl -tag -width Ds -offset indent
.It $ mixer mic.volume=+0.10:-0.05
.El
.Pp
Toggle the mute for
.Ar vol :
.Bl -tag -width Ds -offset indent
.It $ mixer vol.mute=^
.El
.Pp
Set
.Ar mic
and toggle
.Ar line
recording sources:
.Bl -tag -width Ds -offset indent
.It $ mixer mic.recsrc=+ line.recsrc=^
.El
.Pp
Dump
.Ar /dev/mixer0 
information to a file and retrieve back later
.Bl -tag -width Ds -offset indent
.It $ mixer -f /dev/mixer0 -o > info
.It ...
.It $ mixer -f /dev/mixer0 `cat info`
.El
.Sh SEE ALSO
.Xr mixer 3 ,
.Xr sound 4 ,
.Xr sysctl 8
.Sh HISTORY
The
.Nm
utility first appeared in FreeBSD 2.0.5 and was rewritten completely in
FreeBSD 12.2. \" FIXME: replace 12.2 with proper version.
.Sh AUTHORS
.An Christos Margiolis Aq Mt christos@margiolis.net


mixer_prog/mixer_prog.c
/*-
 * Copyright (c) 2021 Christos Margiolis <christos@FreeBSD.org>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * $FreeBSD$
 */

#include <err.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <mixer.h>

static void usage(void) __dead2;
static void initctls(struct mixer *);
static void printall(struct mixer *, int);
static void printminfo(struct mixer *, int);
static void printdev(struct mixer *, int);
static void printrecsrc(struct mixer *, int); /* XXX: change name */
/* Control handlers */
static int mod_dunit(struct mix_dev *, void *);
static int mod_volume(struct mix_dev *, void *);
static int mod_mute(struct mix_dev *, void *);
static int mod_recsrc(struct mix_dev *, void *);
static int print_volume(struct mix_dev *, void *);
static int print_mute(struct mix_dev *, void *);
static int print_recsrc(struct mix_dev *, void *);

static const mix_ctl_t ctl_dunit = {
	.parent_dev	= NULL,
	.id		= -1,
	.name		= "default_unit",
	.mod		= mod_dunit,
	.print		= NULL
};

int
main(int argc, char *argv[])
{
	struct mixer *m;
	mix_ctl_t *cp;
	char *name = NULL, buf[NAME_MAX];
	char *p, *bufp, *devstr, *ctlstr, *valstr = NULL;
	int dunit, i, n, pall = 1;
	int aflag = 0, dflag = 0, oflag = 0, sflag = 0;
	char ch;

	while ((ch = getopt(argc, argv, "ad:f:os")) != -1) {
		switch (ch) {
		case 'a':
			aflag = 1;
			break;
		case 'd':
			dunit = strtol(optarg, NULL, 10);
			if (errno == EINVAL || errno == ERANGE)
				err(1, "strtol");
			dflag = 1;
			break;
		case 'f':
			name = optarg;
			break;
		case 'o':
			oflag = 1;
			break;
		case 's':
			sflag = 1;
			break;
		case '?':
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	/* Print all mixers and exit. */
	if (aflag) {
		if ((n = mixer_get_nmixers()) < 0)
			err(1, "mixer_get_nmixers");
		for (i = 0; i < n; i++) {
			(void)snprintf(buf, sizeof(buf), "/dev/mixer%d", i);
			if ((m = mixer_open(buf)) == NULL)
				err(1, "mixer_open: %s", buf);
			initctls(m);
			if (sflag)
				printrecsrc(m, oflag);
			else {
				printall(m, oflag);
				if (oflag)
					printf("\n");
			}
			(void)mixer_close(m);
		}
		return (0);
	}

	if ((m = mixer_open(name)) == NULL)
		err(1, "mixer_open: %s", name);

	initctls(m);

	if (dflag && ctl_dunit.mod(m->dev, &dunit) < 0)
		goto parse;
	if (sflag) {
		printrecsrc(m, oflag);
		(void)mixer_close(m);
		return (0);
	}

parse:
	while (argc > 0) {
		if ((p = bufp = strdup(*argv)) == NULL)
			err(1, "strdup(%s)", *argv);
		/* Split the string into device, control and value. */
		devstr = strsep(&p, ".");
		if ((m->dev = mixer_get_dev_byname(m, devstr)) == NULL) {
			warnx("%s: no such device", devstr);
			goto next;
		}
		/* Input: `dev`. */
		if (p == NULL) {
			printdev(m, 1);
			pall = 0;
			goto next;
		}
		ctlstr = strsep(&p, "=");
		if ((cp = mixer_get_ctl_byname(m->dev, ctlstr)) == NULL) {
			warnx("%s.%s: no such control", devstr, ctlstr);
			goto next;
		}

		/* Input: `dev.control`. */
		if (p == NULL) {
			(void)cp->print(cp->parent_dev, cp->name);
			pall = 0;
			goto next;
		}
		valstr = p;
		/* Input: `dev.control=val`. */
		cp->mod(cp->parent_dev, valstr);
next:
		free(p);
		argc--;
		argv++;
	}

	if (pall)
		printall(m, oflag);
	(void)mixer_close(m);

	return (0);
}

static void __dead2
usage(void)
{
	printf("usage: %1$s [-f device] [-d unit] [-os] [dev[.control[=value]]] ...\n"
	    "       %1$s [-d unit] [-os] -a\n",
	    getprogname());
	exit(1);
}

static void
initctls(struct mixer *m)
{
	struct mix_dev *dp;
	int rc = 0;

#define C_VOL 0
#define C_MUT 1
#define C_SRC 2
	TAILQ_FOREACH(dp, &m->devs, devs) {
		rc += mixer_add_ctl(dp, C_VOL, "volume", mod_volume, print_volume);
		rc += mixer_add_ctl(dp, C_MUT, "mute", mod_mute, print_mute);
		rc += mixer_add_ctl(dp, C_SRC, "recsrc", mod_recsrc, print_recsrc);
	}
	if (rc) {
		(void)mixer_close(m);
		err(1, "cannot make controls");
	}
}

static void
printall(struct mixer *m, int oflag)
{
	struct mix_dev *dp;

	printminfo(m, oflag);
	TAILQ_FOREACH(dp, &m->devs, devs) {
		m->dev = dp;
		printdev(m, oflag);
	}
}

static void
printminfo(struct mixer *m, int oflag)
{
	int playrec = MIX_MODE_PLAY | MIX_MODE_REC;

	if (oflag)
		return;
	printf("%s: <%s> %s", m->mi.name, m->ci.longname, m->ci.hw_info);
	printf(" (");
	if (m->mode & MIX_MODE_PLAY)
		printf("play");
	if ((m->mode & playrec) == playrec)
		printf("/");
	if (m->mode & MIX_MODE_REC)
		printf("rec");
	printf(")");
	if (m->f_default)
		printf(" (default)");
	printf("\n");
}

static void
printdev(struct mixer *m, int oflag)
{
	struct mix_dev *d = m->dev;
	mix_ctl_t *cp;

	if (!oflag) {
		printf("    %-11s= %.2f:%.2f\t", 
		    d->name, d->vol.left, d->vol.right);
		if (!MIX_ISREC(m, d->devno))
			printf(" pbk");
		if (MIX_ISREC(m, d->devno))
			printf(" rec");
		if (MIX_ISRECSRC(m, d->devno))
			printf(" src");
		if (MIX_ISMUTE(m, d->devno))
			printf(" mute");
		printf("\n");
	} else {
		TAILQ_FOREACH(cp, &d->ctls, ctls) {
			(void)cp->print(cp->parent_dev, cp->name);
		}
	}
}

static void
printrecsrc(struct mixer *m, int oflag)
{
	struct mix_dev *dp;
	int n = 0;

	if (!m->recmask)
		return;
	if (!oflag)
		printf("%s: ", m->mi.name);
	TAILQ_FOREACH(dp, &m->devs, devs) {
		if (MIX_ISRECSRC(m, dp->devno)) {
			if (n++ && !oflag)
				printf(", ");
			printf("%s", dp->name);
			if (oflag)
				printf(".%s=+%s",
				    mixer_get_ctl(dp, C_SRC)->name,
				    n ? " " : "");
		}
	}
	printf("\n");
}

static int
mod_dunit(struct mix_dev *d, void *p)
{
	int dunit = *((int *)p);
	int n;

	if ((n = mixer_get_dunit()) < 0) {
		warn("cannot get default unit");
		return (-1);
	}
	if (mixer_set_dunit(d->parent_mixer, dunit) < 0) {
		warn("cannot set default unit to: %d", dunit);
		return (-1);
	}
	printf("%s: %d -> %d\n", ctl_dunit.name, n, dunit);

	return (0);
}

static int
mod_volume(struct mix_dev *d, void *p)
{
	struct mixer *m;
	mix_ctl_t *cp;
	mix_volume_t v;
	const char *val;
	char lstr[8], rstr[8];
	float lprev, rprev, lrel, rrel;
	int n;

	m = d->parent_mixer;
	cp = mixer_get_ctl(m->dev, C_VOL);
	val = p;
	n = sscanf(val, "%7[^:]:%7s", lstr, rstr);
	if (n == EOF) {
		warnx("invalid volume value: %s", val);
		return (-1);
	}
	lrel = rrel = 0;
	if (n > 0) {
		if (*lstr == '+' || *lstr == '-')
			lrel = rrel = 1;
		v.left = strtof(lstr, NULL);
	}
	if (n > 1) {
		if (*rstr == '+' || *rstr == '-')
			rrel = 1;
		v.right = strtof(rstr, NULL);
	}
	switch (n) {
	case 1:
		v.right = v.left; /* FALLTHROUGH */
	case 2:
		if (lrel)
			v.left += m->dev->vol.left;
		if (rrel)
			v.right += m->dev->vol.right;

		if (v.left < MIX_VOLMIN)
			v.left = MIX_VOLMIN;
		else if (v.left > MIX_VOLMAX)
			v.left = MIX_VOLMAX;
		if (v.right < MIX_VOLMIN)
			v.right = MIX_VOLMIN;
		else if (v.right > MIX_VOLMAX)
			v.right = MIX_VOLMAX;

		lprev = m->dev->vol.left;
		rprev = m->dev->vol.right;
		if (mixer_set_vol(m, v) < 0)
			warn("%s.%s=%.2f:%.2f", 
			    m->dev->name, cp->name, v.left, v.right);
		else
			printf("%s.%s: %.2f:%.2f -> %.2f:%.2f\n",
			   m->dev->name, cp->name, lprev, rprev, v.left, v.right);
	}

	return (0);
}

static int
mod_mute(struct mix_dev *d, void *p)
{
	struct mixer *m;
	mix_ctl_t *cp;
	const char *val;
	int n, opt = -1;

	m = d->parent_mixer;
	cp = mixer_get_ctl(m->dev, C_MUT);
	val = p;
	switch (*val) {
	case '0':
		opt = MIX_UNMUTE;
		break;
	case '1':
		opt = MIX_MUTE;
		break;
	case '^':
		opt = MIX_TOGGLEMUTE;
		break;
	default:
		warnx("%c: no such modifier", *val);
		return (-1);
	}
	n = MIX_ISMUTE(m, m->dev->devno);
	if (mixer_set_mute(m, opt) < 0)
		warn("%s.%s=%c", m->dev->name, cp->name, *val);
	else
		printf("%s.%s: %d -> %d\n",
		    m->dev->name, cp->name, n, MIX_ISMUTE(m, m->dev->devno));

	return (0);
}

static int
mod_recsrc(struct mix_dev *d, void *p)
{
	struct mixer *m;
	mix_ctl_t *cp;
	const char *val;
	int n, opt = -1;

	m = d->parent_mixer;
	cp = mixer_get_ctl(m->dev, C_SRC);
	val = p;
	switch (*val) {
	case '+':
		opt = MIX_ADDRECSRC;
		break;
	case '-':
		opt = MIX_REMOVERECSRC;
		break;
	case '=':
		opt = MIX_SETRECSRC;
		break;
	case '^':
		opt = MIX_TOGGLERECSRC;
		break;
	default:
		warnx("%c: no such modifier", *val);
		return (-1);
	}
	n = MIX_ISRECSRC(m, m->dev->devno);
	if (mixer_mod_recsrc(m, opt) < 0)
		warn("%s.%s=%c", m->dev->name, cp->name, *val);
	else
		printf("%s.%s: %d -> %d\n", 
		    m->dev->name, cp->name, n, MIX_ISRECSRC(m, m->dev->devno));

	return (0);
}

static int
print_volume(struct mix_dev *d, void *p)
{
	struct mixer *m = d->parent_mixer;
	const char *ctl_name = p;

	printf("%s.%s=%.2f:%.2f\n", 
	    m->dev->name, ctl_name, m->dev->vol.left, m->dev->vol.right);

	return (0);
}

static int
print_mute(struct mix_dev *d, void *p)
{
	struct mixer *m = d->parent_mixer;
	const char *ctl_name = p;

	printf("%s.%s=%d\n", m->dev->name, ctl_name, MIX_ISMUTE(m, m->dev->devno));

	return (0);
}

static int
print_recsrc(struct mix_dev *d, void *p)
{
	struct mixer *m = d->parent_mixer;
	const char *ctl_name = p;

	if (!MIX_ISRECSRC(m, m->dev->devno))
		return (-1);
	printf("%s.%s=+\n", m->dev->name, ctl_name);

	return (0);
}
